### DISCUSSION TITLE
Worker mode not working.  What am I doing wrong?

### DISCUSSION THREAD

**Discussion by jamiecarl** [php/frankenphp#1221] [General]
Hi all.  I'm new to FrankenPHP but have my own PHP framework that I'm trying to get working with it.  I have it working fine in non-worker mode and I'm finding it a lot faster than FPM or others.  So now I'm trying to get it working in worker mode and not having much luck.

So what happens is, my worker script runs, starts up and bootstraps my application correctly and I can debug it up to when it executes `frankenphp_handle_request($handler)`, but then that's it.  None of my actual requests make it to the handler.

The confusing part is that everything "appears" to work because requests are being sent to my standard `index.php` entrypoint.

If I remove the `root * /` from the _Caddyfile_ then nothing works and I get 404s.

What am I doing wrong here?  I've been through the documentation a bunch of times and come up blank, but obviously I've missed a step.  I'm hoping someone can shed some light on the issue for me.  Thanks.

Caddy file:

```
{
	frankenphp {
		worker /app/public/worker.php
	}
}

localhost:80 {
	root * public/
	encode zstd br gzip
	php_server
}

```

Worker script:

```php
<?php

declare(strict_types=1);

use Hazaar\Application;

if ((!($_SERVER['FRANKENPHP_WORKER'] ?? false)) || !function_exists('frankenphp_handle_request')) {
    echo 'FrankenPHP must be in worker mode to use this script.';

    exit(1);
}

// Prevent worker script termination when a client connection is interrupted
ignore_user_abort(true);

// Composer autoloading
include __DIR__.'/../vendor/autoload.php';

// Create a new application
$app = new Application(APPLICATION_ENV);
$app->bootstrap(true);

// Handler outside the loop for better performance (doing less work)
$handler = static function () use ($app) {
    $app->run();
};

$requestCount = 0;
$maxRequests = (int) ($_ENV['MAX_REQUESTS'] ?? $_SERVER['MAX_REQUESTS'] ?? 1000);

while ($requestCount < $maxRequests && frankenphp_handle_request($handler)) {
    ++$requestCount;
    gc_collect_cycles();
}

// Cleanup
$app->shutdown();
```

**Comment by jamiecarl** [comment]
Update.  So this is interesting.  If I rename the **worker.php** script file to **index.php** and update the worker config in the Caddyfile _(see below)_, everything works.

???Why???

```
{
	frankenphp {
		worker /app/public/index.php
	}
}

localhost:80 {
	root * public/
	encode zstd br gzip
	php_server
}
```

This isn't mentioned in the documentation, which even has **worker.php** as the name of the script in the examples.

**Reply by dunglas** [reply]
By convention, the path of the worker script must be the path of your front controller. This is how FrankenPHP knows that this request must be handled by the worker script and not the traditional way.

**Reply by jamiecarl** [reply]
Thanks.  Makes sense.  It's just not really clear from the documentation that this is the way it works.  The examples, for example, say this to start a worker using Docker:

```
docker run \
    -e FRANKENPHP_CONFIG="worker /app/path/to/your/worker/script.php" \
    -v $PWD:/app \
    -p 80:80 -p 443:443 -p 443:443/udp \
    dunglas/frankenphp
```

But nowhere does it state that `/app/path/to/your/worker/script.php` needs to be `public/index.php`.  Which my understanding of the default Caddyfile is that it uses `php_server`, which only has `index.php` defined in the `try_files` directive.  So I think that stating that the worker script is `/app/path/to/your/worker/script.php` is a bit misleading seeing as it can really only be `public/index.php`.

But again, I'm new, so let me know if I'm not totally understanding this.

**Reply by dunglas** [reply]
Indeed, we need to clarify the documentation.

**Reply by jamiecarl** [reply]
> Indeed, we need to clarify the documentation.

Cool cool.  On the plus side, once I figured this out, I was able to tweak a few things in my framework (Hazaar MVC) and it runs beautifully in worker mode now.  

Way faster too.  Some quick load time stats on my test page:

PHP-FPM: ~20ms
FrankenPHP: ~8ms
FrankenPHP Worker Mode: ~2ms

I'm pretty freakin' impressed. :)

**Comment by zainsyedz** [comment]
Wait. so  the working directory that we typically use in our org is /var/www/html for the root of the application.

I've just added a few folders here and there to compartmentalize the worker.php file (which isn't in the root of the application but inside the bin/frankenworker/

So, this won't work? I just want to make the code such that it can run traditionally with php-fpm and with our internal framework (which has a different implementation of the index.php) 

I just broke my head for hours till I ran into this discussion.

So in my case I will need to change the root of the application to /app ? 

Can't go with /var/www/html?

**Reply by zainsyedz** [reply]
If I do change the worker to /var/www/html/public/index.php

only if go to localhost/index.php does the handler get  called and the contents of it get executed.

Otherwise, nothing.

This is my index.php (worker)
```
<?php
ignore_user_abort(true);

// Define the public folder root
$publicDir = __DIR__;
$handler = static function () use ($publicDir) {
    var_dump($_SERVER); // Request-bound $_SERVER
    echo "meow";
    return;
};

// Start worker loop
$maxRequests = (int)($_SERVER['MAX_REQUESTS'] ?? 1000);
for ($requestsServed = 0; !$maxRequests || $requestsServed < $maxRequests; ++$requestsServed) {
    $keepRunning = frankenphp_handle_request($handler);
    gc_collect_cycles();
    if (!$keepRunning) break;
}
```

Caddyfile config
```
{
    frankenphp {
        num_threads 4
        max_threads 4
        worker {
            file /var/www/html/public/index.php
            watch /var/www/html/**/*.php
            num 2
        }
    }
    debug
}

localhost:80 {
    encode zstd br gzip
    root * /var/www/html/public
    file_server browse  # Allows directory browsing in dev
    php_server
    log {
        output stdout  # Show logs directly in the container
        level DEBUG
    }
}
```

**Reply by zainsyedz** [reply]
Might I add the app I'm trying to work with worker mode isn't an SPA. 

I guess I have to rewrite the requests to index.php or my worker.php file and then it might start working? 
Somehow preserving the original request uri

**Reply by withinboredom** [reply]
The entrypoint and worker script have to be the same script. Your entrypoint can bootstrap itself like what we do in our integration tests: https://github.com/dunglas/frankenphp/blob/main/testdata/_executor.php; which is used like so: https://github.com/dunglas/frankenphp/blob/main/testdata/cookies.php

**Reply by zainsyedz** [reply]
yeah, I could do that. But then it would do an backward incompatible change to each entrypoint script!

**Reply by zainsyedz** [reply]
I don't know why I thought it would be backward incompatible. I just got it while I was riding on my bike this afternoon! 
Thanks for the tip! 
I was thinking of routing all *.php requests to the worker file for the specific domain name which would work as well. right? This would be done through caddy for each domain. Basically, using rewrite and the original URI can be added with a separate header? 

Your approach would require changing all entrypoints though right, modifying them to return functions so as to make the executor work? This would be backward incompatible. 
Unless I use a similar approach but change it such that it would still work in the traditional sense

**Reply by withinboredom** [reply]
Yes, basically it needs to know how to route the request to the correct thread pool. The `worker` stanza sets up a thread pool, while the route directs the request to the file you want to load. As long as the request matches the thread pool file, it will route to it.

Now that we have worker names, one of the things I want to do is make it so you can simply declare a route’s worker script by name and then have the worker script handle which files to load. It’s on my todo list.

**Reply by zainsyedz** [reply]
i've got no idea about worker names and their role. I just started to delve into FP about a month ago and finally got implementing it in the past week or so! 
Is it part of the frankenphp global option? frankenphp.worker.name? This would in turn be able to allow one to just mention it in the routing specific config? 
So in stead of rewriting it to the specific worker file you can just mention to route all .php files or a specific path to the worker script by just "tagging" it with the worker's defined name?

**Reply by withinboredom** [reply]
basically, something like this:

```
{
	frankenphp {
		worker my_worker /app/public/worker.php
	}
}

localhost:80 {
	root * public/
	encode zstd br gzip
	php_server via my_worker
}
```

or something. I haven't fleshed it out completely yet, but that's the general idea. Basically, it just routes all requests to that route/server to the specified worker.

**Reply by withinboredom** [reply]
There's some refactoring going on in that area, so I'm waiting for it to be done (I haven't checked on it in a bit, TBH).

**Reply by henderkes** [reply]
@AlliBalliBaba and I had a similar brainstorm around simplifying (and optimising) workers a bit. He came up with a php_worker directive (replacing php/php_server), or a match directive in the worker, I thought maybe our existing php/php_server directives could use a match directive.

Similar to how the root matcher in caddy works.

```Caddyfile
php {
    root /path/to/public
    match * worker index.php 4
}

**Reply by AlliBalliBaba** [reply]
Yeah still not sure what the best syntax would be, here are some ideas:

```caddyfile
php_worker worker.php

php_server worker worker.php

php_server {
  match * worker worker.php
}

php_server {
  index worker worker.php
}

php_server {
  index_worker worker.php
}

php_server {
  entrypoint worker.php
}

```

**Reply by zainsyedz** [reply]
``From what I've understood, there should ideally be only one worker (not count, but one worker, that is the primary router) per application per domain.

If that's the case, I think it should be a straightforward syntax where you mention where all php files should be directed too.

For now I've just written one simple rewrite rule for all php files.
There is no nuance here, the request_uri with path remained intact too, so no faffing about there either.

I think php_server via worker_name should be the most straight forward.

Unless you want to have separate workers (example, for api and for dash) for different subdirectories, then we can have a regex pattern option too.

```
php_server {
  match * worker_name
}
```
In that case this makes most sense to me, but rather than mention the actual worker file name, I think just the worker name with match *
